### PR TITLE
Feature/time and weather

### LINKED ISSUE #177: Daylight and Weather cycle
Seeing as this is one of the things in the README roadmap, I'm creating an issue for this with the intention of implementing this sometime soon.

### PR DESCRIPTION
**Pull request by Naamloos**
It's a smaller PR, but it adds some behavior that adds some extra life to the world: world time and weather cycles! 💖

This does not yet include thunderous weather. I'm leaving that for a later PR. closes #177

### CODE REVIEW

File: Obsidian/Client.cs [hunk 1]
```diff
@@ -386,11 +386,18 @@ await this.QueuePacketAsync(new WindowItems(0, this.Player.Inventory.ToList())
...
+        await this.SendTimeUpdateAsync();
+        await this.SendWeatherUpdateAsync();
     }
 
     #region Packet sending
     internal Task DisconnectAsync(ChatMessage reason) => Task.Run(() => SendPacket(new Disconnect(reason, this.State)));
 
+    internal Task SendTimeUpdateAsync() => Task.Run(() => SendPacket(new TimeUpdate(this.Server.World.Data.Time, this.Server.World.Data.DayTime)));
+    internal Task SendWeatherUpdateAsync() => 
+        Task.Run(() => SendPacket(new ChangeGameState(this.Server.World.Data.Raining ? ChangeGameStateReason.BeginRaining : ChangeGameStateReason.EndRaining)));
```
Comment (Seb-stian) [inline Obsidian/Client.cs]: These should probably make use of `QueuePacketAsync` instead of `Task.Run`.

File: Obsidian/WorldData/Level.cs [hunk 2]
```diff
@@ -78,7 +78,18 @@ public class Level
     /// The time of day. 0 is sunrise, 6000 is mid day, 12000 is sunset, 18000 is mid night, 24000 is the next day's 0. 
     /// This value keeps counting past 24000 and does not reset to 0.
```
Comment (Seb-stian) [inline Obsidian/WorldData/Level.cs]: It no longer does after your changes.

File: Obsidian/WorldData/Level.cs [hunk 3]
```diff
@@ -147,12 +158,12 @@ public class Level
     /// <summary>
     /// true if the level is currently experiencing rain, snow, and cloud cover.
     /// </summary>
-    public bool Raining { get; set; }
+    public bool Raining { get; set; } = false; // start a world without rain
```
Comment (Seb-stian) [inline Obsidian/WorldData/Level.cs:160]: Should probably be called `IsRaining`.

File: Obsidian/WorldData/World.cs [hunk 4]
```diff
@@ -279,6 +281,58 @@ public IEnumerable<Entity> GetEntitiesNear(VectorF location, float distance = 10
...
+
+        // Gradually increase and decrease rain levels based on
+        // whether value is in range and what weather is active
+        var oldLevel = this.rainLevel;
+        if (!Data.Raining && this.rainLevel > 0f)
+            this.rainLevel -= 0.01f;
+        else if (Data.Raining && this.rainLevel < 1f)
+            this.rainLevel += 0.01f;
+
+        if(oldLevel != this.rainLevel)
```
Comment (Seb-stian) [inline Obsidian/WorldData/World.cs]: Nit: missing space

File: Obsidian/WorldData/World.cs [hunk 5]
```diff
@@ -279,6 +281,58 @@ public IEnumerable<Entity> GetEntitiesNear(VectorF location, float distance = 10
...
+        {
+            // send new level if updated
+            this.Server.BroadcastPacket(new ChangeGameState(ChangeGameStateReason.RainLevelChange, this.rainLevel));
+            if(rainLevel == 0.2f)
```
Comment (Seb-stian) [inline Obsidian/WorldData/World.cs]: Checking for float equality scares me...
Nit: missing space

File: Obsidian/Server.cs [hunk 6]
```diff
@@ -615,6 +615,8 @@ private async Task LoopAsync()
                 }
             }
 
+            await this.World.DoWorldTickAsync();
```
Comment (Seb-stian) [inline Obsidian/Server.cs:618]: Consider using Server ticks. (in some way?)

File: Obsidian/Client.cs (hunk 1 above)
Comment (Naamloos) [reply Obsidian/Client.cs]: I'll fix this one

File: Obsidian/WorldData/Level.cs (hunk 2 above)
Comment (Naamloos) [reply Obsidian/WorldData/Level.cs]: Will update comment

File: Obsidian/WorldData/Level.cs (hunk 3 above)
Comment (Naamloos) [reply Obsidian/WorldData/Level.cs:160]: `Raining` is consistent with vanilla minecraft, so I will not be changing this one.

File: Obsidian/WorldData/World.cs (hunk 4 above)
Comment (Naamloos) [reply Obsidian/WorldData/World.cs]: ight

File: Obsidian/Server.cs (hunk 6 above)
Comment (Naamloos) [reply Obsidian/Server.cs:618]: Worlds can have independent time in vanilla, so I'd rather keep functionality in the `World` class for future proofing (multi-world)

File: Obsidian/WorldData/World.cs [hunk 7]
```diff
@@ -279,6 +281,60 @@ public IEnumerable<Entity> GetEntitiesNear(VectorF location, float distance = 10
...
+            // amount of ticks in a day is 24000
+            if (Data.Raining)
+            {
+                rainTime = Globals.Random.Next(12000, 24000); // rain lasts 0.5 - 1 day
+            }
+            else
+            {
+                rainTime = Globals.Random.Next(12000, 180000); // clear lasts 0.5 - 7.5 day
+            }
+            Data.RainTime = rainTime;
```
Comment (Jonpro03) [inline Obsidian/WorldData/World.cs:308]: Could we do something like:
```csharp
// Rain lasts 0.5-1 day. Clear lasts 0.5 - 7.5 days.
Data.RainTime = Globals.Random.Next(TICKS_DAY/2, Data.Raining ? TICKS_DAY : (int)(TICKS_DAY * 7.5));
```
or is that too convoluted?

File: Obsidian/WorldData/World.cs [hunk 8]
```diff
@@ -279,6 +281,60 @@ public IEnumerable<Entity> GetEntitiesNear(VectorF location, float distance = 10
...
+                this.Server.BroadcastPacket(new ChangeGameState(this.Data.Raining ? ChangeGameStateReason.BeginRaining : ChangeGameStateReason.EndRaining));
+        }
+
+        if (this.Data.Time % (20 * this.Server.Config.TimeTickSpeedMultiplier) == 0)
+        {
+            // Update client time every second / 20 ticks
+            this.Server.BroadcastPacket(new TimeUpdate(this.Data.Time, this.Data.Time % 24000));
+        }
+
+        await this.ManageChunksAsync();
```
Comment (Jonpro03) [inline Obsidian/WorldData/World.cs:335]: Good idea

File: Obsidian/WorldData/World.cs (hunk 7 above)
Comment (Seb-stian) [reply Obsidian/WorldData/World.cs:308]: I had the same idea, but it indeed is convoluted.

Comment (Jonpro03) [reply Obsidian/WorldData/World.cs:308]: Right. Readability > Concise.

**Review by Jonpro03 — APPROVED** [review]
LGTM

**Review by Seb-stian — APPROVED** [review]

File: Obsidian/WorldData/World.cs (hunk 7 above)
Comment (Naamloos) [reply Obsidian/WorldData/World.cs:308]: I'll think about it. I do like the idea of making it a constant.

Comment (Naamloos) [reply Obsidian/WorldData/World.cs:308]: I'll leave it as-is considering the conversation is resolved

**Pull request merged**